Security - Role Mappings UI

legrego opened the pull request:
# Role Mapping UI 🔐 🗺

Adds a new management interface to allow users with the `manage_security` cluster privilege to manage their role mappings. This UI sits on top of the existing [role mapping API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-put-role-mapping.html).

A role mapping is a rule-based mechanism which tells Elasticsearch how to assign roles to users when authenticating from realms other than `native` and `file`. 

Resolves https://github.com/elastic/kibana/issues/17981


## Licensing
The role mapping interface is available to users with a Gold+ license (i.e., `gold`, `platinum`, `trial`, or `enterprise`). This is because role mappings cannot be used with any of the authentication realms available under the Basic license (`native`, `file`). See the [ES Docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-roles.html#mapping-roles) for more information.



## Features
The role mapping UI supports the following operations:
### View Role Mappings
![image](https://user-images.githubusercontent.com/3493255/71093530-63f44880-2177-11ea-86d0-df556fef5b9c.png)


### Create/Edit Role Mapping
![image](https://user-images.githubusercontent.com/3493255/71093461-3909f480-2177-11ea-87eb-1c2b8254a5dc.png)


### Delete Role Mappings
![image](https://user-images.githubusercontent.com/3493255/71093557-6eaedd80-2177-11ea-9c04-957e46fbbf12.png)


## Exceptional conditions
Role mappings are complex entities, and the UI has several states which attempt to inform users of system conditions which impact their operation:

### Insufficient privileges
Users without the `manage_security` cluster privilege will be presented with an "insufficient privileges" view instead of the actual management interfaces.

![image](https://user-images.githubusercontent.com/3493255/71093648-99009b00-2177-11ea-8eb8-96887229f9aa.png)


**Testing**: Login to Kibana with a user who does not have the `manage_security` cluster privilege, and attempt to manage role mappings


### Incompatible realms
We display a warning message if Elasticsearch does not have any realms enabled which can take advantage of role mappings. This will not prevent role mappings from being managed, but rather makes them unnecessary, as the mappings won't be used.

![image](https://user-images.githubusercontent.com/3493255/71093698-b46ba600-2177-11ea-9bbc-bd08c53b19ed.png)

![image](https://user-images.githubusercontent.com/3493255/71093796-e4b34480-2177-11ea-95db-66630b7c17b1.png)

**Testing**:
- Run with a Gold+ license, but without configuring any additional authentication realms in ES

To verify this message disappears, configure Elasticsearch with an external identity provider, such as:
- LDAP
- SAML
- Kerberos
- PKI
- OIDC


### Inline scripts disabled in Elasticsearch
When inline scripts are disabled in Elasticsearch, then the traditional role templates cannot be used. We display a warning when editing an existing role mapping which contains these templates in this scenario.

![image](https://user-images.githubusercontent.com/3493255/71093885-06143080-2178-11ea-901b-1fd7479688df.png)

**Testing**
1) Create a role mapping which uses the "Role template" template type
2) Stop Elasticsearch, and reconfigure to disable inline scripts by setting: `scripts.allowed_types: ["stored"]` in your `elasticsearch.yml`
3) Restart Elasticsearch, and attempt to edit this role mapping


### Stored scripts disabled in Elasticsearch
When stored scripts are disabled in Elasticsearch, then the "stored script" role templates cannot be used. We display a warning when editing an existing role mapping which contains these templates in this scenario.

![image](https://user-images.githubusercontent.com/3493255/71093947-217f3b80-2178-11ea-82e1-5f2cfec7ddee.png)

**Testing**
1) Create a role mapping which uses the "Stored script" template type
2) Stop Elasticsearch, and reconfigure to disable stored scripts by setting: `scripts.allowed_types: ["inline"]` in your `elasticsearch.yml`
3) Restart Elasticsearch, and attempt to edit this role mapping



## Pending Tasks
- [x] 🎨 Final design review
- [x] 🖊 Copy Review
- [x] 📝 Finish documentation (followup PR)
- [x] 👨‍💻 Code review

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/)
* Commit: 17d77c9d396cd3f97e681d8b053f5f877114b0c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/","number":"16750","commit":"17d77c9d396cd3f97e681d8b053f5f877114b0c3"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/kibana-security (Team:Security)

jportner commented:
Reviewing now!

gchaps commented:
+1 to including the label **Name** label above the text input and the label **Enable Mapping** above the toggle button.

For the Enable Mapping description, how about:
Map roles to users based on their username, groups, and other metadata.  When false, ignore mappings.

legrego commented:
Thanks @ryankeairns and @gchaps -- here's a screenshot with all suggestions applied:

![image](https://user-images.githubusercontent.com/3493255/71825406-cf8c5000-3069-11ea-9bb1-c78af9e640fe.png)

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/)
* Commit: 04309cfc27ec985a5e24635426cfaf5a493041c8

### History
* :green_heart: [Build #16750](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/) succeeded 17d77c9d396cd3f97e681d8b053f5f877114b0c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/","number":"16750","commit":"17d77c9d396cd3f97e681d8b053f5f877114b0c3"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/","number":"18090","commit":"04309cfc27ec985a5e24635426cfaf5a493041c8"}]}
PIPELINE-->

kobelb commented:
ACK: Starting review

legrego commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/edit_role_mapping/components/mapping_info_panel/mapping_info_panel.tsx:
```diff
@@ -0,0 +1,300 @@
+        description={
+          <FormattedMessage
+            id="xpack.security.management.editRoleMapping.roleMappingNameFormRowHelpText"
+            defaultMessage="A distinct name used solely as an identifier."
```
@gchaps any thoughts on Joe's suggestion here?

gchaps replied:
@legrego I agree with Joe's suggestion.

legrego commented:
> When the only non-native / non-file realm enabled is the PKI realm, creating a role mapping with the "groups" field will have no effect (reference). Maybe we can show a warning to the user in that case, and remove the "groups" option from the presets in the user input field?

I'm hesitant to show a warning, as the rules section of the UI is already pretty dense, but I'm not opposed to hiding the "groups" option if we think there's enough benefit. Currently, the features check returns a [simple flag](https://github.com/legrego/kibana/blob/security/role-mappings-ui/x-pack/plugins/security/server/role_mappings/get_enabled_role_mappings_features.ts#L93) indicating if compatible realms are enabled. We'd need to enhance this to return the set of enabled realms instead. This is certainly doable, but comes with a little added complexity to an already complex feature. I'll give this some more thought.

> When creating a brand new role mapping, you can unintentionally overwrite another one by using the same name. Can we prevent the user from doing this, or otherwise warn them when it's about to happen?

For better or worse, the current behavior is consistent with the Roles UI today. The ES APIs for both do not distinguish between create and update, so a pre-flight check would be required on save. Again, this is certainly doable (subject to race conditions), but I was avoiding the additional implementation since we had a precedent for the less-than-ideal behavior 🙈 

> I like that the no_compatible_realms component provides a link to the Elasticsearch documentation for role mapping. I think it would be useful to provide this link when the user has compatible realms, too. It gives more information about role mapping that would be useful for someone who isn't familiar with it. What do you think?

That's not a bad idea -- I do attempt to provide links to more specific doc entries throughout the UI, but the overview might also be helpful. @andreadelrio / @gchaps do you have any thoughts on this?

gchaps commented:
Although I am not opposed to the idea, I worry that we are putting too many links in our UIs.  Where do you plan to add the link? Would it be appropriate to link to an overview topic from the Help menu?

jportner commented:
> Although I am not opposed to the idea, I worry that we are putting too many links in our UIs. Where do you plan to add the link? Would it be appropriate to link to an overview topic from the Help menu?

This is what I had in mind:

![image](https://user-images.githubusercontent.com/5295965/71911636-10f02e80-3142-11ea-83f2-b527c2c362fb.png)

Just a simple link, "Read more here."

gchaps commented:
A link in the intro works.  I would use the text "Learn more."  While I usually prefer the more specific text, I think it is clear that the link goes to the role mapping docs because "role mapping" appears in the title and intro sentence.

jportner commented:
I got a bit confused when seeing this:

![image](https://user-images.githubusercontent.com/5295965/71914895-9676dd00-3148-11ea-98ca-1bd69558b5f3.png)

The two of these rules (`except_all_rule` and `except_field_rule`) are functionally equivalent. It seems that the only reason to include the `except_field_rule` is because the Elasticsearch API supports it, so we have to be able to render that type of rule.

There's already a lot going on in this UI, what do you think about potentially ditching the `except_field_rule` and just auto-converting that to an `except_all_rule` whenever the user is editing an existing role mapping?

I get that the intent of this is to provide a UI for the functionality that Elasticsearch provides, but some things just don't quite make sense when you translate them to a UI, in my opinion.

jportner commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/model/except_field_rule.ts:
```diff
@@ -0,0 +1,82 @@
+    return i18n.translate(
+      'xpack.security.management.editRoleMapping.exceptFieldRule.displayTitle',
+      {
+        defaultMessage: 'Is false',
```
I got a bit confused when seeing this:

![image](https://user-images.githubusercontent.com/5295965/71914895-9676dd00-3148-11ea-98ca-1bd69558b5f3.png)

The two of these rules (`except_all_rule` and `except_field_rule`) are functionally equivalent. It seems that the only reason to include the `except_field_rule` is because the Elasticsearch API supports it, so we have to be able to render that type of rule.

There's already a lot going on in this UI, what do you think about potentially ditching the `except_field_rule` and just auto-converting that to an `except_all_rule` whenever the user is editing an existing role mapping?

I get that the intent of this is to provide a UI for the functionality that Elasticsearch provides, but some things just don't quite make sense when you translate them to a UI, in my opinion.

legrego replied:
> There's already a lot going on in this UI, what do you think about potentially ditching the except_field_rule and just auto-converting that to an except_all_rule whenever the user is editing an existing role mapping?

That's an interesting idea! I do think that'd improve the usability and readability of the rules, but I'm also hesitant to automatically convert it, mostly because we expose the raw rule definition via the JSON editor, so users who are familiar with their rule set would see that the rules they previously created don't exactly match. Even within the same editing session, going from JSON editor -> Visual editor -> JSON editor would illustrate the difference.

I do agree that the current state can get confusing though, and you're absolutely right that they both exist solely because the Elasticsearch API supports it...

Do you think the potential confusion of this rewrite is worth the ux improvement?

jportner replied:
Assuming this would be an easy change like it appears to be -- my personal opinion is that, if such a change was made, the total sum of confusion would be less 😁 
How many people are familiar with role mapping and would notice (or care) if we converted the rule? Probably not many.

legrego replied:
I tested this out: it does appear to be an easy fix and would actually simplify a couple of other code paths slightly. Let's see what others think!

kobelb replied:
I'm also hesitant for us to automatically convert the `except_field_rule` to a `except_any_rule` because of the ability to toggle between JSON and visual editors...

Do we see any similarity to the fact that we allow users to specify the following via the JSON editor:

```
{
  "field": {
    "username": "*"
  }
}
``` 

Which we render as the following, even though the user can't create this rule via just the UI?

<img width="1207" alt="Screen Shot 2020-01-09 at 3 19 48 PM" src="https://user-images.githubusercontent.com/627123/72112778-818a7d00-32f3-11ea-8aa4-53b2fbc48137.png">

legrego replied:
It’s kind of similar, in that the visual editor is limiting what you’re able to do in both cases. The only difference is that the visual editor (and the persisted model) will alter what is on screen if we perform the conversion from except_field_rule. 

Joe does bring up a good point though, in that this is a bit of an edge case that most users won’t run into, let alone notice (we think)

jportner commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/model/except_all_rule.ts:
```diff
@@ -0,0 +1,66 @@
+  /** {@see RuleGroup.getDisplayTitle} */
+  public getDisplayTitle() {
+    return i18n.translate('xpack.security.management.editRoleMapping.exceptAllRule.displayTitle', {
+      defaultMessage: 'Any are false',
```
It seems like this should be swapped with the message in `except_any_rule`?

legrego replied:
hmm, my understanding is the [`except` clause](https://www.elastic.co/guide/en/elasticsearch/reference/current/role-mapping-resources.html#role-mapping-resources) negates the rule that follows. The `all` rule states that _all of the following rules must be true_. Negating that would be _all of the following rules must not be true_, which is how I arrived at `Any are false` for this message.

It's very possible that I'm missing something here though 😄

jportner replied:
Oh, duh, yep.

jportner commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/model/except_any_rule.ts:
```diff
@@ -0,0 +1,71 @@
+  /** {@see RuleGroup.getDisplayTitle} */
+  public getDisplayTitle() {
+    return i18n.translate('xpack.security.management.editRoleMapping.exceptAnyRule.displayTitle', {
+      defaultMessage: 'None are true',
```
Nit: the other messages are "All are true", "Any are true", and "Any are false". I think "All are false" would be more consistent here.

legrego replied:
Good suggestion, changed!

jportner commented:
> > When creating a brand new role mapping, you can unintentionally overwrite another one by using the same name. Can we prevent the user from doing this, or otherwise warn them when it's about to happen?
> 
> For better or worse, the current behavior is consistent with the Roles UI today. The ES APIs for both do not distinguish between create and update, so a pre-flight check would be required on save. Again, this is certainly doable (subject to race conditions), but I was avoiding the additional implementation since we had a precedent for the less-than-ideal behavior 🙈

Yeah, I understand.

I don't think a pre-flight check is necessary, though. You could simply check which role mappings already exist after the `mapping_info_panel` component mounts, and add client-side validation for it. You did [something similar with the `role_selector`](https://github.com/elastic/kibana/pull/53620/files#diff-d16a1df2803793e3c7311be218eabbf2R34-R37) which prevents a user from using a role that doesn't exist.

Just an idea 😁

legrego commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/role_mappings_grid/components/role_mappings_grid_page.tsx:
```diff
@@ -0,0 +1,473 @@
+        field: 'enabled',
+        name: i18n.translate('xpack.security.management.roleMappings.enabledColumnName', {
+          defaultMessage: 'Enabled',
+        }),
```
good idea!

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/)
* Commit: 459a0c3e0e4331282973c52b50856496196594ad

### History
* :broken_heart: [Build #18090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/) failed 04309cfc27ec985a5e24635426cfaf5a493041c8
* :green_heart: [Build #16750](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/) succeeded 17d77c9d396cd3f97e681d8b053f5f877114b0c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/","number":"16750","commit":"17d77c9d396cd3f97e681d8b053f5f877114b0c3"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/","number":"18090","commit":"04309cfc27ec985a5e24635426cfaf5a493041c8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/","number":"18496","commit":"459a0c3e0e4331282973c52b50856496196594ad"}]}
PIPELINE-->

jportner commented on x-pack/plugins/security/server/routes/role_mapping/get.ts:
```diff
@@ -0,0 +1,74 @@
+      if (expectSingleEntity) {
+        if (mappings.length === 0) {
+          return response.notFound();
+        }
```
I can't seem to trigger this code path (e.g., by manually creating a URL to attempt to edit an unknown role mapping). I always get a 500 error. I tinkered with this a bit, I believe you need to wrap the clusterClient call in a try/catch and check the status code to return a 404 error if necessary.

Also could use a test case for this scenario.

legrego replied:
Good catch, I'll revise this!

jportner commented on x-pack/test_utils/stub_web_worker.ts:
```diff
@@ -0,0 +1,16 @@
+    // https://developer.mozilla.org/en-US/docs/Web/API/Worker/terminate
+    this.terminate = jest.fn();
+  };
+}
```
Is this being used? I didn't see anything in your code to indicate that it's using a web worker.
I was wondering if this was perhaps a holdover from an earlier iteration that used a web worker?

legrego replied:
This is being imported into a couple of test suites which end up rendering the JSON rule editor. The underlying `EuiCodeEditor` component uses `brace`/`ace` under the hood, which in turn uses web workers:

Ex:

https://github.com/elastic/kibana/blob/17d77c9d396cd3f97e681d8b053f5f877114b0c3/x-pack/legacy/plugins/security/public/views/management/role_mappings/edit_role_mapping/components/edit_role_mapping_page.test.tsx#L11-L14

jportner replied:
Oh, okay thanks!

jportner commented on x-pack/test/functional/config_security_basic.js:
```diff
@@ -0,0 +1,84 @@
+        '--xpack.maps.showMapsInspectorAdapter=true',
+        '--xpack.maps.preserveDrawingBuffer=true',
+        '--xpack.reporting.queue.pollInterval=3000', // make it explicitly the default
+        '--xpack.reporting.csv.maxSizeBytes=2850', // small-ish limit for cutting off a 1999 byte report
```
Supernit: are all of these config args necessary? Seems like cruft.

I know this was copied from `xpack/test/functional/config.js`, but since the only test file that relies on this is `basic_license`, maybe the extraneous args could be removed. I can just imagine this being a bit confusing in the future.

jportner commented:
I'm done reviewing, everything else LGTM.
I wish this feature was around when I was setting up PKI last month..! This is awesome!

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/)
* Commit: 8899b1cfafca1557e61f762911bead7d9c94a2f6

### History
* :green_heart: [Build #18496](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/) succeeded 459a0c3e0e4331282973c52b50856496196594ad
* :broken_heart: [Build #18090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/) failed 04309cfc27ec985a5e24635426cfaf5a493041c8
* :green_heart: [Build #16750](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/) succeeded 17d77c9d396cd3f97e681d8b053f5f877114b0c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/","number":"16750","commit":"17d77c9d396cd3f97e681d8b053f5f877114b0c3"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/","number":"18090","commit":"04309cfc27ec985a5e24635426cfaf5a493041c8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/","number":"18496","commit":"459a0c3e0e4331282973c52b50856496196594ad"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/","number":"18692","commit":"8899b1cfafca1557e61f762911bead7d9c94a2f6"}]}
PIPELINE-->

legrego commented on x-pack/legacy/plugins/security/common/model.ts:
```diff
@@ -21,3 +21,5 @@ export {
   getUserDisplayName,
 } from '../../../../plugins/security/common/model';
+
+export * from './role_mapping';
```
Nope, I’ll move these to the NP. I started this so long ago that this wasn’t originally the case

legrego commented on x-pack/plugins/security/server/role_mappings/get_enabled_role_mappings_features.ts:
```diff
@@ -0,0 +1,103 @@
+  const xpackUsagePromise: Promise<XPackUsageResponse> = clusterClient
+    .callAsInternalUser('transport.request', {
+      method: 'GET',
+      path: '/_xpack/usage',
```
I didn’t do this originally because it used to be called the shield plugin, and this route isn’t related to security. Are we ok mixing the responsibilities of this ES client plugin?

kobelb replied:
That's a great question that likely warrants further discussion. Our usage of these elasticsearch client plugins needs some rework, and after further investigation there seems to be a rather prolific use of `transport.request` to get around the lack of am "x-pack plugin". Let's not blow the scope up on this PR and we can just leave it the way it is now. Do you mind adding a comment to it about user-input being used to construct the URL?

legrego commented on x-pack/plugins/security/server/role_mappings/get_enabled_role_mappings_features.ts:
```diff
@@ -0,0 +1,103 @@
+    })
+    .catch(error => {
+      logger.error(`Error retrieving node settings for role mappings: ${error}`);
+      return {};
```
I don’t have a specific error in mind here. I was mainly focused on making sure the UI remained usable in the event that this call failed for whatever reason. It’s a value-add feature, and not strictly required for the role mappings UI to function, so my thinking was that we could log that something went wrong, but still let the user continue with their task.

kobelb replied:
Gotcha, so when inline/stored scripts are detected as being disabled we can still create role-mappings which use the inline/stored scripts, they just won't be used? If so, can we add a comment to these try/catch blocks denoting why we have a catch without conditions?

legrego commented on x-pack/plugins/security/server/role_mappings/get_enabled_role_mappings_features.test.ts:
```diff
@@ -0,0 +1,47 @@
+import { getEnabledRoleMappingsFeatures } from './get_enabled_role_mappings_features';
+
+describe('getEnabledRoleMappingsFeatures', () => {
+  // Most scenarios are covered in the `feature_check.test.ts` suite, so will not be repeated here
```
Yeah I can certainly in-line this logic. It started out that way,  but then I split it out because the lengthy route handler looked a bit messy to be, but you’re right, it’s not adding any value standing on its own

kobelb replied:
If you'd like to have it in a separate file, perhaps we could create a folder at `x-pack/plugins/security/server/routes/role_mapping/feature_check` and put this file within that folder?

legrego commented on x-pack/plugins/security/server/role_mappings/get_enabled_role_mappings_features.ts:
```diff
@@ -0,0 +1,103 @@
+  const nodeScriptSettingsPromise: Promise<NodeSettingsResponse> = clusterClient
+    .callAsInternalUser('transport.request', {
+      method: 'GET',
+      path: '/_nodes/settings?filter_path=nodes.*.settings.script',
```
I’m pretty comfortable with this approach. I was given this suggestion by the ES team on Slack when I  told them what I was trying to accomplish

kobelb replied:
Good enough for me then :)

legrego commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/role_mappings_grid/components/create_role_mapping_button/create_role_mapping_button.tsx:
```diff
@@ -0,0 +1,25 @@
+  return (
+    <EuiButton
+      data-test-subj="createRoleMappingButton"
+      onClick={() => (window.location.hash = EDIT_ROLE_MAPPING_PATH)}
```
Oops, great catch, yes I can certainly change this to use href instead

legrego commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/edit_role_mapping/services/role_mapping_validation.ts:
```diff
@@ -0,0 +1,93 @@
+  return valid();
+}
+
+export function validateRoleMappingRoleTemplates({
```
Yeah ES allows this, so I decided to allow the UI to do so as well...

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/)
* Commit: 6272cec1e5d22dbe2b1a2ff384c4626325412dc7

### History
* :broken_heart: [Build #18692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/) failed 8899b1cfafca1557e61f762911bead7d9c94a2f6
* :green_heart: [Build #18496](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/) succeeded 459a0c3e0e4331282973c52b50856496196594ad
* :broken_heart: [Build #18090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/) failed 04309cfc27ec985a5e24635426cfaf5a493041c8
* :green_heart: [Build #16750](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/) succeeded 17d77c9d396cd3f97e681d8b053f5f877114b0c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/","number":"16750","commit":"17d77c9d396cd3f97e681d8b053f5f877114b0c3"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/","number":"18090","commit":"04309cfc27ec985a5e24635426cfaf5a493041c8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/","number":"18496","commit":"459a0c3e0e4331282973c52b50856496196594ad"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/","number":"18692","commit":"8899b1cfafca1557e61f762911bead7d9c94a2f6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/","number":"18986","commit":"6272cec1e5d22dbe2b1a2ff384c4626325412dc7"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/)
* Commit: 47c95efd333e90b1a4684b7f53dfc657aeae0f40

### History
* :green_heart: [Build #18986](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/) succeeded 6272cec1e5d22dbe2b1a2ff384c4626325412dc7
* :broken_heart: [Build #18692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/) failed 8899b1cfafca1557e61f762911bead7d9c94a2f6
* :green_heart: [Build #18496](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/) succeeded 459a0c3e0e4331282973c52b50856496196594ad
* :broken_heart: [Build #18090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/) failed 04309cfc27ec985a5e24635426cfaf5a493041c8
* :green_heart: [Build #16750](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/) succeeded 17d77c9d396cd3f97e681d8b053f5f877114b0c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16750/","number":"16750","commit":"17d77c9d396cd3f97e681d8b053f5f877114b0c3"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/","number":"18090","commit":"04309cfc27ec985a5e24635426cfaf5a493041c8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/","number":"18496","commit":"459a0c3e0e4331282973c52b50856496196594ad"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/","number":"18692","commit":"8899b1cfafca1557e61f762911bead7d9c94a2f6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/","number":"18986","commit":"6272cec1e5d22dbe2b1a2ff384c4626325412dc7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/","number":"19090","commit":"47c95efd333e90b1a4684b7f53dfc657aeae0f40"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/)
* Commit: c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54

### History
* :green_heart: [Build #19090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/) succeeded 47c95efd333e90b1a4684b7f53dfc657aeae0f40
* :green_heart: [Build #18986](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/) succeeded 6272cec1e5d22dbe2b1a2ff384c4626325412dc7
* :broken_heart: [Build #18692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/) failed 8899b1cfafca1557e61f762911bead7d9c94a2f6
* :green_heart: [Build #18496](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/) succeeded 459a0c3e0e4331282973c52b50856496196594ad
* :broken_heart: [Build #18090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/) failed 04309cfc27ec985a5e24635426cfaf5a493041c8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18090/","number":"18090","commit":"04309cfc27ec985a5e24635426cfaf5a493041c8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/","number":"18496","commit":"459a0c3e0e4331282973c52b50856496196594ad"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/","number":"18692","commit":"8899b1cfafca1557e61f762911bead7d9c94a2f6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/","number":"18986","commit":"6272cec1e5d22dbe2b1a2ff384c4626325412dc7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/","number":"19090","commit":"47c95efd333e90b1a4684b7f53dfc657aeae0f40"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/","number":"19105","commit":"c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54"}]}
PIPELINE-->

kobelb commented on x-pack/plugins/security/server/routes/role_mapping/post.test.ts:
```diff
@@ -0,0 +1,100 @@
+import { defineRoleMappingPostRoutes } from './post';
+
+describe('POST role mappings', () => {
+  it('allows a role mapping to be created', async () => {
```
🙈

kobelb commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/edit_role_mapping/services/role_mapping_validation.ts:
```diff
@@ -0,0 +1,93 @@
+}
+
+export function validateRoleMappingRoles({ roles }: RoleMapping): ValidationResult {
+  if (roles && !roles.length) {
```
Gotcha, thank for the clarification!

legrego commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19268/)
* Commit: c4bb9d38bdf0f95ae7172b625829a39b85e0e56e

### History
* :broken_heart: [Build #19105](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/) failed c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54
* :green_heart: [Build #19090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/) succeeded 47c95efd333e90b1a4684b7f53dfc657aeae0f40
* :green_heart: [Build #18986](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/) succeeded 6272cec1e5d22dbe2b1a2ff384c4626325412dc7
* :broken_heart: [Build #18692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/) failed 8899b1cfafca1557e61f762911bead7d9c94a2f6
* :green_heart: [Build #18496](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/) succeeded 459a0c3e0e4331282973c52b50856496196594ad

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18496/","number":"18496","commit":"459a0c3e0e4331282973c52b50856496196594ad"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/","number":"18692","commit":"8899b1cfafca1557e61f762911bead7d9c94a2f6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/","number":"18986","commit":"6272cec1e5d22dbe2b1a2ff384c4626325412dc7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/","number":"19090","commit":"47c95efd333e90b1a4684b7f53dfc657aeae0f40"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/","number":"19105","commit":"c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19268/","number":"19268","commit":"c4bb9d38bdf0f95ae7172b625829a39b85e0e56e"}]}
PIPELINE-->

legrego commented on x-pack/legacy/plugins/security/public/views/management/role_mappings/model/except_field_rule.ts:
```diff
@@ -0,0 +1,82 @@
+/**
+ * Represents a negated field rule {@link FieldRule}.
+ */
+export class ExceptFieldRule extends RuleGroup {
```
I _think_ I'm leaning towards Joe's suggestion, despite my concerns. Thankfully, none of these options have to be permanent. We can change our minds in a followup, and it won't be a breaking change or involve a ton of work or refactoring

kobelb replied:
Cool, let's do it then! I don't think there is a *right* answer here

legrego replied:
Done in https://github.com/elastic/kibana/pull/53620/commits/041099e63953e437e3192782c2c657a5dd437cb0!

![image](https://user-images.githubusercontent.com/3493255/72169409-109f9f80-339d-11ea-985e-eae87de62c03.png)

legrego commented:
@gchaps as we discussed, I added "infomercial" style docs to this PR, linking to the existing ES docs where appropriate

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19329/)
* Commit: a6918e38a5ead8ed8a933ef39cca049da67d651d

### History
* :green_heart: [Build #19268](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19268/) succeeded c4bb9d38bdf0f95ae7172b625829a39b85e0e56e
* :broken_heart: [Build #19105](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/) failed c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54
* :green_heart: [Build #19090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/) succeeded 47c95efd333e90b1a4684b7f53dfc657aeae0f40
* :green_heart: [Build #18986](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/) succeeded 6272cec1e5d22dbe2b1a2ff384c4626325412dc7
* :broken_heart: [Build #18692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/) failed 8899b1cfafca1557e61f762911bead7d9c94a2f6

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18692/","number":"18692","commit":"8899b1cfafca1557e61f762911bead7d9c94a2f6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/","number":"18986","commit":"6272cec1e5d22dbe2b1a2ff384c4626325412dc7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/","number":"19090","commit":"47c95efd333e90b1a4684b7f53dfc657aeae0f40"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/","number":"19105","commit":"c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19268/","number":"19268","commit":"c4bb9d38bdf0f95ae7172b625829a39b85e0e56e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19329/","number":"19329","commit":"a6918e38a5ead8ed8a933ef39cca049da67d651d"}]}
PIPELINE-->

legrego commented:
@kobelb I _think_ I addressed all of your feedback (🤞), so this is ready for another review whenever you are. I tried to reference commit hashes in my replies to your comments, so it's easier to find/review in isolation

kobelb commented:
ACK: re-reviewing now

legrego commented:
I'm going to merge ahead of docs approval to unblock other work, so I'll address any feedback in a followup PR.

@jportner your username is attached to a bunch of new snyk checks on this PR now, one of which is failing due to a missing manifest. Looks unrelated to any of the work in this PR, so I'm ignoring.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19531/)
* Commit: ab20b177c440233b232f8d1673be6421cc4c0bfc

### History
* :green_heart: [Build #19329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19329/) succeeded a6918e38a5ead8ed8a933ef39cca049da67d651d
* :green_heart: [Build #19268](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19268/) succeeded c4bb9d38bdf0f95ae7172b625829a39b85e0e56e
* :broken_heart: [Build #19105](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/) failed c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54
* :green_heart: [Build #19090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/) succeeded 47c95efd333e90b1a4684b7f53dfc657aeae0f40
* :green_heart: [Build #18986](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/) succeeded 6272cec1e5d22dbe2b1a2ff384c4626325412dc7

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/18986/","number":"18986","commit":"6272cec1e5d22dbe2b1a2ff384c4626325412dc7"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/","number":"19090","commit":"47c95efd333e90b1a4684b7f53dfc657aeae0f40"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/","number":"19105","commit":"c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19268/","number":"19268","commit":"c4bb9d38bdf0f95ae7172b625829a39b85e0e56e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19329/","number":"19329","commit":"a6918e38a5ead8ed8a933ef39cca049da67d651d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19531/","number":"19531","commit":"ab20b177c440233b232f8d1673be6421cc4c0bfc"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/25980/)
* Commit: ab20b177c440233b232f8d1673be6421cc4c0bfc

### History
* :green_heart: [Build #19531](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19531/) succeeded ab20b177c440233b232f8d1673be6421cc4c0bfc
* :green_heart: [Build #19329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19329/) succeeded a6918e38a5ead8ed8a933ef39cca049da67d651d
* :green_heart: [Build #19268](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19268/) succeeded c4bb9d38bdf0f95ae7172b625829a39b85e0e56e
* :broken_heart: [Build #19105](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/) failed c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54
* :green_heart: [Build #19090](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/) succeeded 47c95efd333e90b1a4684b7f53dfc657aeae0f40

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19090/","number":"19090","commit":"47c95efd333e90b1a4684b7f53dfc657aeae0f40"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19105/","number":"19105","commit":"c8d21b6dddcca97e4f5a3f85e538c7c8da1efe54"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19268/","number":"19268","commit":"c4bb9d38bdf0f95ae7172b625829a39b85e0e56e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19329/","number":"19329","commit":"a6918e38a5ead8ed8a933ef39cca049da67d651d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19531/","number":"19531","commit":"ab20b177c440233b232f8d1673be6421cc4c0bfc"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/25980/","number":"25980","commit":"ab20b177c440233b232f8d1673be6421cc4c0bfc"}]}
PIPELINE-->
